webextension/privileged/ui/CloudDownloadsView.jsm

```diff
@@ -0,0 +1,739 @@
+  },
+
+  _addRadioOption(key, providerName, document) {
+    const option = document.createElement("radio");
+    option.id = key;
+    option.type = "radio";
+
+    option.setAttribute("class", "panelCloudNotificationUI-radio");
+    option.setAttribute("label", "Save to " + providerName);
+    option.setAttribute("provider", this._formatProviderName(providerName));
+    return option;
+  },
+
+  _addNotificationMultipleProviders(providersMap, document) {
+    const multiProviderSelect = document.getElementById("multiProviderSelect");
+
+    const providers = [];
+    providersMap.forEach((value, key) => {
+      providers.push({"key": key, "name": value.displayName});
+    });
+
+    // Reset multiProviderSelect by removing any previous
+    // added options
+    while (multiProviderSelect.firstChild) {
+      multiProviderSelect.firstChild.remove();
+    }
+
+    for (const provider of providers) {
+      multiProviderSelect.appendChild(
+        this._addRadioOption(provider.key, provider.name, document));
+    }
+    multiProviderSelect.removeAttribute("hidden");
+  },
+
+  _iconURL(name) {
+    if (this._formatProviderName(name) === "localdownloads") {
+      return "moz-icon://" + CloudDownloadsInternal.defaultDownloadDirIconURL + "?size=16";
+    }
+    return new URL(this.stylesURL).origin + "/skin/" + this._formatProviderName(name) + ".svg";
+  },
+
+  _setMenuItemAttributes(menuItem, providerKey, providerDetail, downloadElement) {
+    menuItem.setAttribute("label", providerDetail.displayName);
+    menuItem.setAttribute("providerKey", providerKey);
+    menuItem.setAttribute("class", "menuitem-iconic");
+    menuItem.setAttribute("image", this._iconURL(providerDetail.displayName));
+    return menuItem;
+  },
+
+  _setMoveDownloadMenuPopUpAttributes(menuItem) {
+    menuItem.removeAttribute("hidden");
+    const menuPopup = menuItem.menupopup;
+    menuPopup.removeAttribute("hidden");
+    return menuPopup.firstChild;
+  },
+
+  _displayMoveToCloudContextMenuItem(downloadElement) {
+    const aPopupMenu = downloadElement.ownerDocument.getElementById("downloadsContextMenu");
+    let menuItem = aPopupMenu.getElementsByAttribute("id", "moveDownload")[0];
+    if (menuItem) {
+      menuItem.setAttribute("hidden", "true");
+
+      // Exit if user has opted to store subsequent downloads in cloud provider folder
+      if (CloudDownloadsInternal.preferredProviderKey) {
+        return;
+      }
+
+      if (CloudDownloadsInternal.checkIfExistingCloudProviderDownloadSettings()) {
+        return;
+      }
+
+      const downloadType = downloadElement.getAttribute("cloudstorage");
+      // Downloaded Item is saved in local downloads folder
+      if (this.providers.size > 1) {
+        let subMenuItem =  this._setMoveDownloadMenuPopUpAttributes(menuItem);
+        this.providers.forEach((value, key) => {
+          if (this._formatProviderName(value.displayName) !== downloadType) {
```

Comment:
For local downloads, the downloadType will be "local". However, the formatted provider name for a local provider is "localdownloads". We should settle on 1 key. Right now I expect there will be subtle bugs where we check against the "wrong" value. Obviously on this particular line that can't be the case because `this.providers` will never include the "local" case...

Reply:
done